### PR TITLE
refactor: try to use select in async stream function

### LINKED ISSUE #4947: refactor: `QueryResultFetcher` should add a shutdown channel to listen on task failure
Now, after all stages are scheduled, `QueryResultFetcher` will begin to work and use `get_data` to fetch results. 

According to #3908 , the executor can not throw error through channels. It's possible that `QueryResultFetcher` may already start to work (All stages are created), but some other task get execution failed later, in this case the `QueryResultFetcher` won't know it.

One solution is to add a shutdown channel channel when running `QueryResultFetcher`, if one task execution fail, this shutdown channel will receive the info and return error.

cc @liurenjie1024

### PR DESCRIPTION
**Pull request by BowenXiao1999**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Close #4947 

However, seems not easy to add a tokio::select in futures-async-stream.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/frontend/src/scheduler/distributed/query_manager.rs [hunk 1]
```diff
@@ -132,6 +144,32 @@ impl QueryResultFetcher {
             yield DataChunk::from_protobuf(response?.get_record_batch()?)?;
         }
     }
+
+    fn run(self) -> BoxedDataChunkStream {
+        Box::pin(self.run_inner())
+    }
+
+    // #[try_stream(ok = DataChunk, error = RwError)]
+    async fn run_wrapper(self, shutdown_rx: Receiver<u8>) -> impl DataChunkStream {
```
Comment (BowenXiao1999) [inline src/frontend/src/scheduler/distributed/query_manager.rs:153]: Tried to build a wrapper on original function and use async-stream's stream! macro.

**Comment by BowenXiao1999** [comment]
There are some other potential solutions:
1. Manually use if to `try_recv` from the channel. 
2. Manually implement the #[try_stream] just like https://crates.io/crates/futures-async-stream. (Just like `RemoteInput`, `LocalInput`, needs investigation). 

cc @BugenZhao

File: src/frontend/src/scheduler/distributed/query_manager.rs [hunk 2]
```diff
@@ -132,6 +144,32 @@ impl QueryResultFetcher {
...
+        let mut shutdown_rx = shutdown_rx;
+        let mut stream = self.run();
+        stream! {
```
Comment (BugenZhao) [inline src/frontend/src/scheduler/distributed/query_manager.rs:156]: Does
```rust
self.run_inner().take_until(shutdown_rx)
```
work for you?

File: src/frontend/src/scheduler/distributed/query.rs [hunk 3]
```diff
@@ -137,7 +138,7 @@ impl QueryExecution {
     }
 
     /// Start execution of this query.
-    pub async fn start(&self) -> SchedulerResult<QueryResultFetcher> {
+    pub async fn start(&self, shutdown_tx: Sender<u8>) -> SchedulerResult<QueryResultFetcher> {
```
Comment (BugenZhao) [inline src/frontend/src/scheduler/distributed/query.rs:141]:
```suggestion
    pub async fn start(&self, shutdown_tx: Sender<()>) -> SchedulerResult<QueryResultFetcher> {
```

File: src/frontend/src/scheduler/distributed/query_manager.rs (hunk 2 above)
Comment (BowenXiao1999) [reply src/frontend/src/scheduler/distributed/query_manager.rs:156]: I‘ll try

Comment (BowenXiao1999) [reply src/frontend/src/scheduler/distributed/query_manager.rs:156]: This take_until works very well!

But there brings another question: I can only terminate the stream early, but not yield Error inside the Stream. This means even if the shutdown_rx trigger, the calller who consume can not know error happend but just know the stream stopped. 

TakeUntil has a method named `take_result`, so the solution may be: after consume the stream, use `take_result` to check stopping future resolved or not. If yes, extract the error inside it and returned. But now we are using BoxedStream or impl Stream every where, now all need to change to `TakeUntil<'_>` 😢

File: src/frontend/src/handler/query.rs [hunk 4]
```diff
@@ -54,17 +56,28 @@ pub async fn handle_query(
...
         rows.extend(to_pg_rows(chunk?, format));
     }
 
+    // Check whether error happen, if yes, returned.
+    let execution_ret = data_stream.take_result();
+    if let Some(ret) = execution_ret {
+        return Err(ret
+            .expect("The shutdown message receiver should not fail")
+            .into());
+    }
```
Comment (BowenXiao1999) [inline src/frontend/src/handler/query.rs:79]: I choose `take_until` + `take_result` to early terminate a Stream, and checked correspoding trigger error.

**Review by liurenjie1024 — APPROVED** [review]
LGTM

**Pull request merged**